### ISSUE TITLE
installing an older version of x86_64 pulls in a newer version of i686 through dependencies

### ISSUE THREAD

**Issue by lukash** [openSUSE/libsolv#404]
I've discovered this while constructing a test case for #383. The test data are basically the same, I just replaced `libstoragemgmt` with `foo` to make the names shorter.

The test case is here (if you prefer me to aggregate the test case into a single piece of text and put it inline here, let me know): https://github.com/lukash/libsolv/tree/multilib-install

The interesting thing here is [testcase.t](https://github.com/lukash/libsolv/blob/multilib-install/test/testcases/multilib-install/testcase.t) tries to install oneof `foo-1.0-1.i686 foo-1.0-1.x86_64`, which results in correctly installing `foo-1.0-1.x86_64`, but `foo-2.0-1.i686` gets installed as well ([solver.result](https://github.com/lukash/libsolv/blob/multilib-install/test/testcases/multilib-install/solver.result)), because it appears when resolving the dependencies a newer version is picked along the way, and then to satisfy that, i686 in a newer version is the only solution...

**Comment by mlschroe** [comment]
That's not really a bug. It goes like this:

foo-1.0-1.i686 requires python3-foo provided by python3-foo-1.0-1.x86_64 and python3-foo-2.0-1.x86_64. The solver picks the higher version.

python3-foo-2.0-1.x86_64 requires foo = 2.0-1, provided by foo-2.0-1.i686 and foo-2.0-1.x86_64. It cannot install the x86_64 package because it conflicts with foo-1.0, so it has to pick the i686 version.

So you end up with both the i686 and x86_64 version of foo. Those different architectures act as if the packages have different names.

Things would be different if you drop that "setarch" flag from the job, as in that case the solver can only install the i686 package if the corresponding x86_64 package gets installed as well. (I questioned dnf's use of setarch in a couple of other issues as well...)

The underlying issue is that "setarch" disables the lockstep for all package of that name, and not only the packages from the job. I.e. it also disables the lockstep of the foo-2.0 package even if your job just addresses foo-1.0 packages. Maybe I can change this.

**Comment by lukash** [comment]
> That's not really a bug. It goes like this:
> 
> ...

Right, I've figured that's basically what's happening. I understand the mechanism, but it is certainly unexpected.

> Things would be different if you drop that "setarch" flag from the job, as in that case the solver can only install the i686 package if the corresponding x86_64 package gets installed as well. (I questioned dnf's use of setarch in a couple of other issues as well...)

We know, I've done some investigation and we had a discussion about it. The result is that we don't know what the implications of removing the flag would be, as we just don't understand the vast underlying logic in libsolv and are just unable to evaluate the effect. Since this area of depsolving has a massive amount of corner-cases that aren't covered in our test suite, we could easily introduce a major breakage somewhere and we don't want to risk that :slightly_frowning_face: At the moment we just don't have the capacity to dive into libsolv to figure this all out. If you have any suggestion at all how to approach this, it would be very welcome :slightly_smiling_face: 

> The underlying issue is that "setarch" disables the lockstep for all package of that name, and not only the packages from the job. I.e. it also disables the lockstep of the foo-2.0 package even if your job just addresses foo-1.0 packages. Maybe I can change this.

Can you please elaborate on what you mean by "lockstep"?
